### PR TITLE
Eng 2082 add metrics to `get_artifact_versions` route

### PR DESCRIPTION
**Pull request by likawind**
## Describe your changes and why you are making these changes
This PR adds metrics to `get_artifact_versions` route. We also moved around a few codes in this route to make the code more readable. I will add in-line comments explaining changes.

## Related issue number (if any)
ENG-2082

## Demo
```
{
    "latest_versions": {
        "cf625895-475b-41bf-a150-a918aa22338e": {
            "workflow_name": "test_conda",
            "artifact_name": "copy_more artifact",
            "workflow_id": "26ebf1b9-8156-4385-b4aa-51e527d8771b",
            "workflow_dag_result_id": "00000000-0000-0000-0000-000000000000",
            "artifact_id": "cf625895-475b-41bf-a150-a918aa22338e",
            "load_specs": [
                {
                    "service": "SQLite",
                    "integration_id": "2bef3c43-9ee8-4e46-a909-612d68aace97",
                    "parameters": {
                        "table": "review_copied",
                        "update_mode": "replace"
                    }
                }
            ],
            "versions": {
                "18376563-64d1-49ef-9e75-e5bb00652fe6": {
                    "timestamp": 1670612218,
                    "status": "succeeded",
                    "error": "",
                    "checks": null,
                    "metrics": [
                        {
                            "id": "5181f974-94a8-4d2c-93f2-9bf9d5136a06",
                            "name": "row_count artifact",
                            "description": "",
                            "type": "numeric",
                            "from": "00000000-0000-0000-0000-000000000000",
                            "to": null,
                            "result": {
                                "id": "0fe7c374-f0f0-4737-83f5-db2114015ecf",
                                "serialization_type": "json",
                                "content_path": "7ab9863b-dae4-4630-9ddc-5250bcd6ec3d",
                                "content_serialized": "100",
                                "exec_state": {
                                    "user_logs": {
                                        "stdout": "",
                                        "stderr": ""
                                    },
                                    "status": "succeeded",
                                    "failure_type": null,
                                    "error": null,
                                    "timestamps": {
                                        "registered_at": null,
                                        "pending_at": "2022-12-09T18:56:58.227667795Z",
                                        "running_at": "2022-12-09T18:57:12.270909092Z",
                                        "finished_at": "2022-12-09T18:57:16.698961079Z"
                                    }
                                }
                            }
                        }
                    ]
                }
            }
        }
    },
    "historical_versions": {}
}
```

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 1]
```diff
@@ -76,21 +80,25 @@ func (*GetArtifactVersionsHandler) Prepare(r *http.Request) (interface{}, int, e
 	return aqContext, http.StatusOK, nil
 }
 
-func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs interface{}) (interface{}, int, error) {
-	args := interfaceArgs.(*aq_context.AqContext)
-
-	emptyResponse := getArtifactVersionsResponse{}
-
+func (h *GetArtifactVersionsHandler) initializeLatestAndHistoricalVersions(
```
Comment (likawind) [inline src/golang/cmd/server/handler/get_artifact_versions.go:83]: this is a helper function just to group some existing codes in `Perform`

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 2]
```diff
@@ -124,15 +129,15 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
...
-		return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to get artifact versions.")
+		return nil, nil, nil, errors.Wrap(err, "Unable to get artifact versions.")
 	}
 
 	latestArtifactIDs := make(map[uuid.UUID]bool, len(artifactIDs))
 	for _, artifactID := range artifactIDs {
 		latestArtifactIDs[artifactID] = true
 	}
 
-	allArtifactIds := make([]uuid.UUID, 0, len(loadOperatorIDs))
```
Comment (likawind) [inline src/golang/cmd/server/handler/get_artifact_versions.go:135]: naming convention

File: src/golang/cmd/server/handler/get_artifact_versions.go (hunk 1 above)
Comment (likawind) [reply src/golang/cmd/server/handler/get_artifact_versions.go:83]: No behavior change within this function

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 3]
```diff
@@ -200,23 +219,32 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
...
-	for failedArtifactId := range failedArtifactIdsMap {
+	failedArtifactIDs := make([]uuid.UUID, 0, len(failedArtifactIDsMap))
+	for failedArtifactId := range failedArtifactIDsMap {
 		failedArtifactIDs = append(failedArtifactIDs, failedArtifactId)
 	}
 
+	return failedArtifactIDs, failedDAGResultIDs, nil
+}
+
+func (h *GetArtifactVersionsHandler) updateVersionsWithChecksAndMetrics(
```
Comment (likawind) [inline src/golang/cmd/server/handler/get_artifact_versions.go:235]: Helper function with original checks code and newly added metrics code

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 4]
```diff
@@ -239,34 +267,163 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
...
-	if len(failedArtifactIDs) > 0 {
-		failedStatuses, err := h.OperatorResultRepo.GetStatusByDAGResultAndArtifactBatch(
-			ctx,
-			failedDAGResultIDs,
-			failedArtifactIDs,
-			h.Database,
-		)
-		if err != nil {
-			return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to get artifact versions.")
+	metricsByUpstreamArtifactID, err := h.ArtifactRepo.GetMetricsByUpstreamArtifactBatch(ctx, allArtifactIDs, h.Database)
```
Comment (likawind) [inline src/golang/cmd/server/handler/get_artifact_versions.go:270]: These are main changes related to metrics

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 5]
```diff
@@ -239,34 +267,163 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
...
+					historicalVersions[upstreamArtifactID].Versions[metricResult.DAGResultID] = artifactVersionObject
 				}
+			}
+		}
+	}
+	return nil
+}
 
-				historicalVersions[failedStatus.ArtifactID].Versions[failedStatus.DAGResultID] = artifactVersionObject
+func (h *GetArtifactVersionsHandler) updateVersionsWithErrorMessages(
```
Comment (likawind) [inline src/golang/cmd/server/handler/get_artifact_versions.go:351]: Another helper to organize existing codes. No behavior change

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 6]
```diff
@@ -239,34 +267,163 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
...
 			}
+
+			historicalVersions[failedStatus.ArtifactID].Versions[failedStatus.DAGResultID] = artifactVersionObject
+		}
+	}
+
+	return nil
+}
+
+func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs interface{}) (interface{}, int, error) {
```
Comment (likawind) [inline src/golang/cmd/server/handler/get_artifact_versions.go:389]: `Perform` is much simplified now

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 7]
```diff
@@ -172,14 +180,25 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
 		}
 	}
 
-	artifactResultStatuses, err := h.ArtifactResultRepo.GetStatusByArtifactBatch(ctx, allArtifactIds, h.Database)
+	return latestVersions, historicalVersions, allArtifactIDs, nil
+}
+
+func (h *GetArtifactVersionsHandler) updateVersionsWithArtifactResultStatues(
```
[the remark this answers is not in the document]
Comment (agiron123) [reply src/golang/cmd/server/handler/get_artifact_versions.go:186]: @likawind Should the name here be updateVersionsWithArtifactResultStatuses? Statues are different than statuses :) 🗽

**Review by agiron123 — APPROVED** [review]

File: src/golang/lib/repos/artifact.go [hunk 8]
```diff
@@ -37,6 +37,15 @@ type artifactReader interface {
 		DB database.Database,
 	) ([]uuid.UUID, error)
 
+	// GetMetricsBatchByUpstreamArtifactBatch returns a map of metrics Artifacts if they
```
Comment (cw75) [inline src/golang/lib/repos/artifact.go:40]: `GetMetricsByUpstreamArtifactBatch`

File: src/golang/lib/repos/sqlite/artifact_result.go [hunk 9]
```diff
@@ -120,8 +120,10 @@ func (*artifactResultReader) GetStatusByArtifactBatch(
 	query := fmt.Sprintf(
 		`SELECT 
 			artifact_result.artifact_id, 
+			artifact_result.id as artifact_result_id,
 			artifact_result.workflow_dag_result_id, 
-			artifact_result.status, 
+			artifact_result.status,
```
Comment (cw75) [inline src/golang/lib/repos/sqlite/artifact_result.go:125]: I think I removed this in the other PR because it's unused.

**Review by cw75 — APPROVED** [review]
Couple nits otherwise LGTM!

File: src/golang/cmd/server/handler/get_artifact_versions.go [hunk 10]
```diff
@@ -239,34 +267,163 @@ func (h *GetArtifactVersionsHandler) Perform(ctx context.Context, interfaceArgs
...
+			metricResults, ok := metricResultsByArtfID[metric.ID]
+			if !ok {
+				continue
+			}
+
+			for _, metricResult := range metricResults {
+				var contentPtr *string = nil
+				dag, ok := dagsByMetricResultID[metricResult.ID]
+				if ok {
+					storageObj := storage.NewStorage(&dag.StorageConfig)
```
Comment (cw75) [inline src/golang/cmd/server/handler/get_artifact_versions.go:315]: Seems that this line should go inside `if metric.Type.IsCompact()`.

**Pull request merged**